gst/droidcodec/gstdroidvdec.c

```diff
@@ -333,21 +334,55 @@ gst_droidvdec_convert_buffer (GstDroidVDec * dec,
       }
     }
   } else {
-    /* copy to the output buffer swapping the u and v planes and cropping if necessary */
-    /* this assumes a NV12 format with 128 byte alignment */
-    gint stride = ALIGN_SIZE (width, 128);
-    gint slice_height = ALIGN_SIZE (height, 32);
-    gint top = ALIGN_SIZE (dec->crop_rect.top, 2);
-    gint left = ALIGN_SIZE (dec->crop_rect.left, 2);
-
-    guint8 *y = in->data + (top * stride) + left;
-    guint8 *uv = in->data + (stride * slice_height) + (top * stride / 2) + left;
-
-    gst_droidvec_copy_plane (map_info.data + info->offset[0], info->stride[0],
-        y, stride, info->width, info->height);
-    gst_droidvec_copy_packed_planes (map_info.data + info->offset[1],
-        map_info.data + info->offset[2], info->stride[1], uv, stride,
-        info->width / 2, info->height / 2);
+    DroidMediaColourFormatConstants c;
+    droid_media_colour_format_constants_init (&c);
+    if (dec->hal_format == c.OMX_COLOR_FormatYUV420Planar) {
+      /* Buffer is already I420, so we can copy it straight over */
+      /* though we need to handle the cropping */
```

Comment:
Theoretically you should be able to attach the crop information to the buffer or the format when it is negotiated and something further down the pipeline like ideally the renderer would take care of the cropping.

Reply:
Yes this would be much better. I'll look into doing this, as maybe there is other metadata I should be passing between the buffers too.

Reply:
While the cropping worked, the buffer needs fixing up too because the codec overrides the dimensions, at which point it's probably easier to just the cropping at the same time. Will merge this as is.